src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/PosixSignalRegistration.Unix.cs

```diff
@@ -0,0 +1,173 @@
+                {
+                    if (signalRegistrations.Count != 0)
+                    {
+                        var registrations = new PosixSignalRegistration?[signalRegistrations.Count];
+                        bool hasRegistrations = false;
+                        for (int i = 0; i < signalRegistrations.Count; i++)
+                        {
+                            if (signalRegistrations[i].TryGetTarget(out PosixSignalRegistration? registration))
+                            {
+                                registrations[i] = registration;
+                                hasRegistrations = true;
+                            }
+                        }
+                        if (hasRegistrations)
+                        {
+                            // This is called on the native signal handling thread. We need to move to another thread so
+                            // signal handling is not blocked. Otherwise we may get deadlocked when the handler depends
+                            // on work triggered from the signal handling thread.
+                            // We use a new thread rather than queueing to the ThreadPool in order to prioritize handling
+                            // in case the ThreadPool is saturated.
+                            Thread handlerThread = new Thread(HandleSignal)
+                            {
+                                IsBackground = true,
+                                Name = ".NET Signal Handler"
+                            };
+                            handlerThread.Start((signal, registrations));
+                            return 1;
+                        }
+                    }
+                }
+            }
+            return 0;
+        }
+
+        private static void HandleSignal(object? state)
+        {
+            var (signal, registrations) = ((PosixSignal, PosixSignalRegistration?[]))state!;
+
+            PosixSignalContext ctx = new(signal);
+            foreach (var registration in registrations)
+            {
+                registration?.Handle(ctx);
+            }
+
+            if (!ctx.Cancel)
+            {
+                Interop.Sys.HandlePosixSignal(signal);
+            }
+        }
+
+        ~PosixSignalRegistration()
+            => Dispose(false);
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (_registered)
+            {
+                lock (s_registrations)
+                {
+                    List<WeakReference<PosixSignalRegistration>> signalRegistrations = s_registrations[_signal];
+                    for (int i = 0; i < signalRegistrations.Count; i++)
+                    {
+                        if (signalRegistrations[i].TryGetTarget(out PosixSignalRegistration? registration) &&
+                            object.ReferenceEquals(this, registration))
```

Comment:
In the case of finalization, won't TryGetTarget return false as the weak reference will have already be cleaned up?  In that case, what removes the WeakReference from the list?

Reply:
Good point. I need to add an `else` branch here.